// @sap/cds-compiler/lib/base/messages.js
// Functions and classes for syntax messages

'use strict';

// For messageIds, where no severity has been provided via code (central def)
const standardSeverities = {
  'syntax-anno-after-struct': ['Error'],
  'syntax-anno-after-enum': ['Error'],
  'syntax-anno-after-params': ['Error'],
  'ref-undefined-def': 'Error',
  'ref-undefined-art': 'Error',
  'anno-undefined-def': 'Info', // for annotate statement (for CSN or CDL path cont)
  'anno-undefined-art': 'Info', // for annotate statement (for CDL path root)
  'anno-undefined-element': 'Info',
  'anno-undefined-action': 'Info',
  'anno-undefined-param': 'Info',
}

// For messageIds, where no text has been provided via code (central def)
const standardTexts = {
  'syntax-anno-after-struct': 'Avoid annotation assignments after structure definitions',
  'syntax-anno-after-enum': 'Avoid annotation assignments after enum definitions',
  'syntax-anno-after-params': 'Avoid annotation assignments after parameters',
  'ref-undefined-def': {
    std: 'Artifact $(ART) has not been found',
    element: 'Artifact $(ART) has no element $(MEMBER)'
  },
  'ref-undefined-art': 'No artifact has been found with name $(NAME)',
  'ref-undefined-element': {
    std: 'Element $(ART) has not been found',
    element: 'Artifact $(ART) has no element $(MEMBER)'
  },
  'anno-undefined-def': 'Artifact $(ART) has not been found',
  'anno-undefined-art': 'No artifact has been found with name $(NAME)',
  'anno-undefined-element': {
    std: 'Element $(ART) has not been found',
    element: 'Artifact $(ART) has no element $(MEMBER)',
    enum: 'Artifact $(ART) has no enum $(MEMBER)'
  },
  'anno-undefined-action': {
    std: 'Action $(ART) has not been found',
    action: 'Artifact $(ART) has no action $(MEMBER)'
  },
  'anno-undefined-param': {
    std: 'Parameter $(ART) has not been found',
    param: 'Artifact $(ART) has no parameter $(MEMBER)'
  },
  'expected-const': 'A constant value is expected here',
  'expected-struct': 'A structured type or a non-query entity without parameters is expected here',
  'expected-context': 'A context or service is expected here',
  'expected-type': 'A type or an element of a type is expected here',
  'expected-entity': 'An entity, projection or view is expected here',
  'expected-source': 'A query source must be an entity or an association to an entity',
}

function hasErrors( messages ) {
  return messages && messages.some( m => m.severity === 'Error' );
}

// Return gnu-style error string for location `loc`:
//  - 'File:Line:Col' without `loc.end`
//  - 'File:Line:StartCol-EndCol' if Line = start.line = end.line
//  - 'File:StartLine.StartCol-EndLine.EndCol' otherwise
function locationString( loc, normalizeFilename ) {
  if (!loc)
    return '<???>';
  let filename = (loc.filename && (normalizeFilename || normalizeFilename === 0))
      ? loc.filename.replace( /\\/g, '/' )
      : loc.filename;
  if (!(loc instanceof Object) || !loc.start)
    return loc;
  return (!loc.end)
    ? `${filename}:${loc.start.line}:${loc.start.column}`
    : (loc.start.line == loc.end.line)
    ? `${filename}:${loc.start.line}:${loc.start.column}-${loc.end.column}`
    : `${filename}:${loc.start.line}.${loc.start.column}-${loc.end.line}.${loc.end.column}`;
}

// Class for combined compiler errors.  Additional members:
//  `errors`: vector of errors (CompileMessage and errors from peg.js)
//  `model`: the CSN model
// TODO: standard param order
class CompilationError extends Error {
  constructor(errs, model, text, ...args) {
    super( text || 'CDS compilation failed\n' + errs.map( m => m.toString() ).join('\n'),
           ...args );
    this.errors = errs;         // TODO: rename to messages
    Object.defineProperty( this, 'model', { value: model, configurable: true } );
  }
  toString() {                  // does not really help -> set message
    return this.message.includes('\n')
      ? this.message
      : this.message + '\n' + this.errors.map( m => m.toString() ).join('\n');
  }
}

// Class for individual compile errors.
// TODO: make it really extend Error?, change param order
class CompileMessage extends Error {
  constructor(location, msg, severity = 'Error', id, home) {
    super(msg);
    this.location = location;
    if (home)                   // semantic location, e.g. 'entity:"E"/element:"x"'
      this.home   = home;
    this.severity = severity;
    if (id)
      Object.defineProperty( this, 'messageId', { value: id } );
      // this.messageId = id;  // ids not yet finalized
  }
  toString() {                  // should have no argument...
    return messageString( this, undefined, true ); // no message-id before finalization!
  }
}

function handleMessages( model ) {
  if (model.messages && model.messages.length) {
    model.messages.sort( compareMessage );
    if (hasErrors( model.messages ))
      throw new CompilationError( model.messages, model );
  }
}

const severitySpecs = {
  error: { name: 'Error', level: 0 },
  warning: { name: 'Warning', level: 1 },
  info: { name: 'Info', level: 2 },
  debug: { name: 'Debug', level: 3 }
}

function normalizedSeverity( severity ) {
  if (typeof severity !== 'string')
    return (severity === null) ? 'Debug' : 'Error';
  let s = severitySpecs[ severity.toLowerCase() ];
  return s && s.name || 'Error';
}

function getMessageFunction( model ) {
  if (!model.messages)
    model.messages = [];
  let config = model.options && model.options.severities || {};

  return function message( id, location, home, params = {}, severity = undefined, texts = undefined ) {
    if (!severity)              // TODO: check that they are always eq per messageId
      severity = standardSeverities[id];
    let s = normalizedSeverity( severity );
    if ((s !== 'Error' || severity instanceof Array) && id && id in config )
      s = normalizedSeverity( config[id] );
    let text = (typeof params === 'string')
        ? params
        : messageText( texts || standardTexts[id], params );
    let msg = new CompileMessage( location, text, s, id,
                                  (typeof home === 'string' ? home : homeName(home)) );
    model.messages.push( msg );
    return msg;
  }
}

const paramsTransform = {
  alias: quoted,
  anno: transformAnno,
  art: transformArg,
  target: transformArg,
  token: t => t.match( /^[a-zA-Z]+$/ ) ? t.toUpperCase() : "'" + t + "'",
  code: n => '`' + n + '`',
  newcode: n => '`' + n + '`',
  name: quoted,
  names: transformManyWith( quoted ),
  id: quoted,
  file: s => "'" + s.replace( /'/g, "''" ) + "'", // sync ;
};

function transformAnno( anno ) {
  return (anno.charAt() === '@') ? quoted( anno ) : quoted( '@' + anno );
  // if (anno.charAt() === '@')
  //   anno = anno.slice(1);
  // return (!anno || /[^A-Za-z_0-9.]/.test(anno)) ? msgName( '@' + anno ) : '@' + anno;
}

function transformArg( arg, r, args, texts ) {
  if (!arg || typeof arg !== 'object')
    return quoted( arg );
  if (args['#'] || args.member )
    return artName( arg );
  if (arg._artifact)
    arg = arg._artifact;
  let name = arg.name;
  if (!name)
    return quoted( name );
  let prop = ['element','param','action','alias'].find( p => name[p] );
  if (!prop || !texts[prop] )
    return artName( arg );
  r['#'] = texts[ name.$variant ] && name.$variant || prop; // text variant (set by searchName)
  r.member = quoted( name[prop] );
  return artName( arg, prop );
}

function transformManyWith( t ) {
  return function transformMany( many, r, args, texts ) {
    let prop = ['none','one'][ many.length ];
    if (!prop || !texts[prop] || args['#'] )
      return many.map(t).join(', ');
    r['#'] = prop;              // text variant
    return many.length && t( many[0] );
  };
}

const nameProp = {
  enum: 'element',
  key: 'element',
  function: 'action',
};

function searchName( art, id, variant ) {
  if (!variant) {
    let type = art._finalType && art._finalType.kind !== 'undefined' ? art._finalType : art;
    art = type.target && type.target._artifact || type;
    variant = ['context','service','namespace'].includes(art.kind) ? 'absolute' : 'element';
  }
  let prop = nameProp[variant] || variant;
  let name = Object.assign( { $variant: variant }, (art._artifact||art).name );
  name[prop] = (name[prop]) ? name[prop] + '.' + id : id || '?';
  return { name, kind: art.kind };
}

function messageText( texts, params ) {
  if (typeof texts === 'string')
    texts = { std: texts };
  let args = {};
  for (let p in params) {
    let t = paramsTransform[p];
    args[p] = (t) ? t( params[p], args, params, texts ) : params[p];
  }
  let variant = args['#'];
  return replaceInString( variant && texts[ variant ] || texts.std, args );
}

function replaceInString( text, params ) {
  let pattern = /\$\(([A-Z]+)\)/g;
  let parts = [];
  let start = 0;
  for (let p = pattern.exec( text ); p; p = pattern.exec( text )) {
    let prop = p[1].toLowerCase();
    parts.push( text.substring( start, p.index ),
                (prop in params ? params[prop] : p[0]) );
    delete params[prop];
    start = pattern.lastIndex;
  }
  parts.push( text.substring( start ) );
  let remain = ('#' in params) ? [] : Object.keys( params ).filter( n => params[n] );
  return (remain.length)
         ? parts.join('') + '; ' +
           remain.map( n => n.toUpperCase() + ' = ' + params[n] ).join(', ')
         : parts.join('');
}

function weakLocation( loc ) {
  // use return { ...location, $weak: true } if that is JS standard
  return { filename: loc.filename, start: loc.start, end: loc.end, $weak: true };
}

// Return message string with location if present
function messageString( err, normalizeFilename, noMessageId, noHome ) {
  return (err.location ? locationString( err.location, normalizeFilename ) + ': ' : '') +
         (err.severity||'Error') +
         (err.messageId && !noMessageId ? ' ' + err.messageId + ': ' : ': ') +
         err.message +
         (!err.home || noHome && err.location && !err.location.$weak ? '' : ' (in ' + err.home + ')');
}

// Compare two messages `a` and `b`. Return 0 if they are equal, 1 if `a` is
// greater than `b`, and -1 if `a` is less than `b`. Messages without a location
// are considered equal, unless compared to a message with location, in which
// case they are considered larger, so they are put at the end of a sorted list
function compareMessage( a, b ) {
  if (a.location && b.location) {
    let aend = a.location.end || a.location.start;
    let bend = b.location.end || b.location.start;
    return ( c( a.location.filename, b.location.filename ) ||
             c( a.location.start.line, b.location.start.line ) ||
             c( a.location.start.column, b.location.start.column ) ||
             c( aend.line, bend.line ) ||
             c( aend.column, bend.column ) ||
             c( a.message, b.message ) );
  }
  else
    return (!a.location ? (!b.location ? 0 : 1) : -1);

  function c( x, y ) {
    return (x == y) ? 0 : (x > y) ? 1 : -1;
  }
}

function artName( art, omit ) {
  let name = art.name;
  let r = (name.absolute) ? [ quoted( name.absolute ) ] : [];
  if (name.query || art.kind === 'block') // Yes, omit $query.0 - TODO: rename to block
    r.push( (art.kind === 'block' ? 'block:' : 'query:') + name.query );
  if (name.action && omit !== 'action')
    r.push( memberActionName(art) + ':' + quoted( name.action ) );
  if (name.param && omit !== 'param') // TODO: also use for alias/mixin
    r.push( 'param:' + quoted( name.param ) );
  else if (name.alias)          // TODO: use 'param'
    r.push( (name.$mixin ? 'mixin:' : 'alias:') + quoted( name.alias ) )
  if (name.element && omit !== 'element')
    r.push( (art.kind === 'enum' ? 'enum:' : 'element:') + quoted( name.element ) );
  return r.join('/');
}

function memberActionName( art ) {
  while (art && art._main) {
    if (art.kind === 'action' || art.kind === 'function')
      return art.kind;
    art = art._parent;
  }
  return 'action';
}

function homeName( art ) {
  if (!art)
    return art;
  if (art._outer)               // in returns / items property
    return homeName( art._outer );
  else if (art.kind === 'source' || !art.name) // error reported in parser or on source level
    return null;
  else if (art.kind === 'using')
    return 'using:' + quoted( art.name.id );
  else if (art.name._artifact)             // block, extend, annotate
    return homeName( art.name._artifact ); // use corresponding definition
  else
    return (art._main ? art._main.kind : art.kind) + ':' + artName( art );
}

function quoted( name ) {
  return (name) ? '"' + name.replace( /"/g, '""' ) + '"' : '<?>'; // sync ";
}

module.exports = {
  hasErrors,
  weakLocation,
  locationString,
  messageString,
  searchName,
  getMessageFunction,
  artName,
  handleMessages,
  sortMessages: (m => m.sort(compareMessage)),
  CompileMessage,
  CompilationError
}
