// @sap/cds-sql/lib/index.js
const cds = require('./cds')

const dependencies = {
  get BaseClient () {
    const BaseClient = require('./client/BaseClient')
    Object.defineProperty(dependencies, 'BaseClient', { value: BaseClient })
    return BaseClient
  },
  get errors () {
    const errors = require('./errors')
    Object.defineProperty(dependencies, 'errors', { value: errors })
    return errors
  },
  get builder () {
    const sqlBuilder = require('./sql-builder/')
    Object.defineProperty(dependencies, 'builder', { value: sqlBuilder })
    return sqlBuilder
  },
  get expand () {
    const expand = require('./expand/expandCQNToJoin')
    expand.rawToExpanded = require('./expand/rawToExpanded')

    Object.defineProperty(dependencies, 'expand', { value: expand })
    return expand
  },
  get composition () {
    const composition = require('./composition/compositionTree')
    Object.defineProperty(dependencies, 'composition', { value: composition })
    return composition
  },
  get contains () {
    const contains = require('./contains/containsCQNToLike')

    Object.defineProperty(dependencies, 'contains', { value: contains })
    return contains
  },
  get postProcessing () {
    const postProcessing = require('./data-conversion/post-processing')
    Object.defineProperty(dependencies, 'postProcessing', { value: postProcessing })
    return postProcessing
  },
  get onCond () {
    const onCond = require('./oncond/generateOnCond')
    Object.defineProperty(dependencies, 'getOnCond', { value: onCond })
    return onCond
  },
  inject: (...args) => {
    return cds.inject(...args)
  }
}

module.exports = dependencies
