// @sap/cds/bin/utils/term.js
const debug = process.env.DEBUG

// https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences
const term = {
  reset: '\x1b[0m', // Default
  bold: '\x1b[1m', // Bold/Bright
  red: '\x1b[91m', // Bright Foreground Red
  green: '\x1b[32m', // Foreground Green
  orange: '\x1b[38;2;255;140;0m' // darker orange, works with bright and dark background
}

const asErr = module.exports.error = o => debug ? o : as(term.red + term.bold, o)
module.exports.warn  = o => as(term.orange, o)
module.exports.info  = o => as(term.green, o)
Object.defineProperty(module.exports, 'colors', { get: () => process.stdin.isTTY && process.stdout.isTTY })

// 'Error: foo'  ->  '[ERROR] foo'  (Maven-like, allows for better grepping in logs)
module.exports.decorate = {
  error: o => debug ? o : `[${asErr('ERROR')}] ${stripErrPrefix(o)}`
}

function as(codes, o) {
  return module.exports.colors ? (codes + o + term.reset) : ('' + o)
}

// strips the general 'Error: ' prefix, so that we can add our own
function stripErrPrefix(o) {
  return o && o.toString ? o.toString().replace(/^Error: /i, '') : o
}
