// @sap/cds-ql/lib/cds.js
const _notInjected = () => {
  throw new Error('This method requires `cds` to be injected')
}

const _validateInjection = cds => {
  if (!cds || typeof cds !== 'object') {
    throw new Error('Injected value is not of type `cds`')
  }
}

const injection = {
  cds: {
    parse: {
      cql: _notInjected,
      expr: _notInjected
    },
    load: _notInjected,
    linked: _notInjected
  },
  inject (cds) {
    _validateInjection(cds)

    if (Object.keys(cds).length === 0) {
      return
    }

    require('@sap/cds-sql').inject(cds)

    injection.cds = cds
  },
  parse: {
    expr (arg) {
      return injection.cds.parse.expr(arg)
    },
    cql (arg) {
      return injection.cds.parse.cql(arg)
    }
  },
  load (...args) {
    return injection.cds.load(...args)
  },
  unfold: {
    for: {
      odata (...args) {
        return injection.cds.unfold.for.odata(...args)
      }
    }
  },
  get env () {
    return injection.cds.env || {}
  },
  get config () {
    return injection.cds.config || {}
  },
  linked (...args) {
    return injection.cds.linked(...args)
  }
}

module.exports = injection
