// @sap/odata-commons/lib/http/HttpStatusCode.js
'use strict';

module.exports = {
    StatusCodes: {
        OK: 200,
        CREATED: 201,
        ACCEPTED: 202,
        NO_CONTENT: 204,
        NOT_MODIFIED: 304,
        BAD_REQUEST: 400,
        UNAUTHORIZED: 401,
        NOT_FOUND: 404,
        METHOD_NOT_ALLOWED: 405,
        NOT_ACCEPTABLE: 406,
        CONFLICT: 409,
        PRECONDITION_FAILED: 412,
        UNSUPPORTED_MEDIA_TYPE: 415,
        FAILED_DEPENDENCY: 424,
        PRECONDITION_REQUIRED: 428,
        INTERNAL_SERVER_ERROR: 500,
        NOT_IMPLEMENTED: 501
    },
    Texts: {
        200: 'OK',
        201: 'Created',
        202: 'Accepted',
        204: 'No Content',
        304: 'Not Modified',
        400: 'Bad Request',
        401: 'Unauthorized',
        404: 'Not Found',
        405: 'Method Not Allowed',
        406: 'Not Acceptable',
        409: 'Conflict',
        412: 'Precondition Failed',
        415: 'Unsupported Media Type',
        424: 'Failed Dependency',
        428: 'Precondition Required',
        500: 'Internal Server Error',
        501: 'Not Implemented'
    }
};
