// @sap/generator-cds/lib/constants.js
const constants = {

    ENV_KEY_CDS_IDE: '_CDS_IDE',
    ENV_KEY_CDS_IDE_WEBIDE: 'webide',
    ENV_KEY_CDS_IDE_ECLIPSE: 'eclipse',

    ENV_VAR_GENERATION_LOG: 'GENERATION_LOG',

    // add new module type here
    MODULE_TYPE_DB: 'db',
    MODULE_TYPE_SRV: 'srv',

    DB_TECHNOLOGY_HANA: 'hana',
    PRIMARY_DB: 'db',
    HDI_CONTAINER_TYPE: 'com.sap.xs.hdi-container',

    SRV_TECHNOLOGY_JAVA: 'java',
    SRV_TECHNOLOGY_NODEJS: 'nodejs',

    DEFAULT_GEN_FOLDER: 'gen',

    DEFAULT_MODULE_FOLDER: {},

    TASKS: {}

};

// add new module type here
constants.MODULE_TYPES = [constants.MODULE_TYPE_DB, constants.MODULE_TYPE_SRV];

// add new module type here
constants.TASKS[constants.MODULE_TYPE_DB] = {};
constants.TASKS[constants.MODULE_TYPE_DB][constants.DB_TECHNOLOGY_HANA] = 'hana';

constants.TASKS[constants.MODULE_TYPE_SRV] = {};
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_JAVA] = 'java-cf';
constants.TASKS[constants.MODULE_TYPE_SRV][constants.SRV_TECHNOLOGY_NODEJS] = 'node-cf';

// add new module type here
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_DB] = 'db';
constants.DEFAULT_MODULE_FOLDER[constants.MODULE_TYPE_SRV] = 'srv';

const regex = {
    REGEX_DB_TECHNOLOGY: `^(${constants.DB_TECHNOLOGY_HANA})?$`,
    REGEX_MODULE_TYPE: `(${constants.MODULE_TYPES.join('|')})`
}


module.exports = {
    ...constants,
    ...regex,
    REGEX_MODULES: `^(${regex.REGEX_MODULE_TYPE}(,${regex.REGEX_MODULE_TYPE})*)?$`,
    REGEX_SRV_TECHNOLOGY: `(${constants.SRV_TECHNOLOGY_NODEJS}|${constants.SRV_TECHNOLOGY_JAVA})`
};
