// @sap/cds-compiler/lib/compiler/builtins.js
// The builtin artifacts of CDS

'use strict';

const { forEachInDict } = require('../base/dictionaries');

var core = {
  String: { parameters: ['length'] },
  LargeString: {},
  Binary: { parameters: ['length'] },
  LargeBinary: {},
  Decimal: { parameters: ['precision', 'scale'] },
  DecimalFloat: {},
  Integer64: {},
  Integer: {},
  Double: {},
  Date: {},
  Time: {},
  DateTime: {},
  Timestamp: {},
  Boolean: {},
  Association: {internal:true},
  Composition: {internal:true},
  UUID: {},
};

var hanaHana = {
  ALPHANUM: { parameters: ['length'] },
  SMALLINT: {},
  TINYINT: {},
  SMALLDECIMAL: {},
  REAL: {},
  CHAR: { parameters: ['length'] },
  NCHAR: { parameters: ['length'] },
  VARCHAR: { parameters: ['length'] },
  CLOB: {},
  BINARY: { parameters: ['length'] },
  ST_POINT: { parameters: [ {name: 'length', literal: 'number', val: 0} ]},
  ST_GEOMETRY: { parameters: [ {name: 'length', literal: 'number', val: 0} ] }
}

var hana = {
  BinaryFloat: {},
  LocalDate: {},
  LocalTime: {},
  UTCDateTime: {},
  UTCTimestamp: {},
  WithStructuredPrivilegeCheck: { kind: 'annotation' },
  hana: { kind: 'context' }
}

const magicVariables = {        // in SQL-92
  CURRENT_DATE: {},
  CURRENT_TIME: {},
  CURRENT_TIMESTAMP: {},
  CURRENT_USER: {},
  SESSION_USER: {},
  // SQL-92: also SYSTEM_USER, just USER (is with parens in HANA SQL), VALUE
  '$user': {
    elements: { id: {}, locale: {} },
    $autoElement: 'id'
  },                  // CDS-specific, not part of SQL
  '$now': {},                   // Dito
}

const magicVariables_hana = {
  CURRENT_CONNECTION: {},
  CURRENT_SCHEMA: {},
  CURRENT_TRANSACTION_ISOLATION_LEVEL: {},
  CURRENT_UTCDATE: {},
  CURRENT_UTCTIME: {},
  CURRENT_UTCTIMESTAMP: {},
  SYSUUID: {},
}

function initBuiltins( model ) {
  if (!model.options.hanaFlavor) {
    setMagicVariables( magicVariables );
    let cds = {
      kind: 'namespace',
      name: { absolute: 'cds' },
      blocks: [],
      artifacts: Object.create(null) };
    // setProp( model.definitions, 'cds', cds );
    model.definitions.cds = cds; // not setProp - oData
    model.$builtins = env( core, 'cds.', undefined, cds );
    model.$builtins.cds = cds;
  }
  else {
    setMagicVariables( Object.assign( {}, magicVariables, magicVariables_hana ) );
    let artifacts = env( Object.assign( {}, core, hana ), 'cds.', 'sap.cds::' );
    artifacts.hana.artifacts = env( hanaHana, 'hana.' );
    model.$builtins = artifacts;
  }
  model.$internal = { $frontend: '$internal' };
  return;

  function env( builtins, prefix, extraPrefix, parent ) {
    let artifacts = Object.create(null);
    for (let name in builtins) {
      let absolute = prefix + name;
      let art = { kind: 'type', builtin: true, name: {absolute}, type: { path:[{id:absolute}] } };
      setProp( art.type, '_artifact', art );
      if (parent)
        parent.artifacts[ name ] = art;
      setProp( art, '_finalType', art );
      setProp( art, '_deps', [] );
      Object.assign( art, builtins[name] );
      if (!art.internal)
        artifacts[name] = art;
      setProp( model.definitions, absolute, art );
      if (extraPrefix)
        setProp( model.definitions, extraPrefix + name, art );
    }
    return artifacts;
  }

  function setMagicVariables( builtins ) {
    let artifacts = Object.create(null);
    for (let name in builtins) {
      let magic = builtins[name];
      // TODO: rename to $builtinFunction
      let art = { kind: 'builtin', name: { id: name, element: name } };
      artifacts[name] = art;
      if (magic.elements)
        art.elements = forEachInDict( magic.elements, (e,n) => magicElement( e, n, art ));
      if (magic.$autoElement)
        art.$autoElement = magic.$autoElement;
      // setProp( art, '_finalType', art );
    }
    model.$magicVariables = { kind: '$magicVariables', artifacts };
  }

  function magicElement( spec, name, parent ) {
    let magic = {
      kind: 'builtin',
      name: { id: name, element: parent.name.element + '.' + name }
    };
    setProp( magic, '_parent', parent );
    // setProp( magic, '_finalType', magic );
    return magic;
  }
}

// Like `obj.prop = value`, but not contained in JSON / CSN
function setProp ( obj, prop, value ) {
  Object.defineProperty( obj, prop, { value, configurable: true, writable: true } ); // not enumerable!
  return value;
}

module.exports = initBuiltins;
