// @sap/cds-services/lib/services/handlers/index.js
/**
 * @module handlers
 */
module.exports = {
  beforeAll: require('./beforeAll'),
  beforeCreate: require('./beforeCreate'),
  beforeCreateDraft: require('./beforeCreateDraft'),
  beforeDeleteDraft: require('./beforeDeleteDraft'),
  beforeEntityAuth: require('./beforeEntityAuth'),
  beforeUpdate: require('./beforeUpdate'),
  beforeUpdateDraft: require('./beforeUpdateDraft'),
  finalOn: require('./finalOn'),
  onCommit: require('./onCommit'),
  onCreate: require('./onCreate'),
  onCreateDraft: require('./onCreateDraft'),
  onDelete: require('./onDelete'),
  onDeleteDraft: require('./onDeleteDraft'),
  onDraftActivate: require('./onDraftActivate'),
  onDraftActivateEvent: require('./onDraftActivateEvent'),
  onDraftEdit: require('./onDraftEdit'),
  onDraftPrepare: require('./onDraftPrepare'),
  onRead: require('./onRead'),
  onReadDraft: require('./onReadDraft'),
  onReadOverDraft: require('./onReadOverDraft'),
  onRollback: require('./onRollback'),
  onUpdate: require('./onUpdate'),
  onUpdateDraft: require('./onUpdateDraft')
}
